In [149]:
#Import scikit-learn dataset library
from sklearn import datasets

# Import train_test_split function
from sklearn.model_selection import train_test_split

#Import Gaussian Naive Bayes model
from sklearn.naive_bayes import GaussianNB

#Import scikit-learn metrics module for accuracy calculation
from sklearn import metrics
import random as rd

from sklearn.feature_extraction.text import CountVectorizer

import pandas as pd

*The* data is now in the three folders `easy_ham`, `hard_ham`, and `spam`.

In [3]:
!ls -lah

total 7696
drwxr-xr-x    14 wushmachine  staff   448B Apr 24 10:09 .
drwxr-xr-x@   11 wushmachine  staff   352B Mar 27 17:37 ..
-rw-r--r--@    1 wushmachine  staff   6.0K Apr 24 10:09 .DS_Store
drwxr-xr-x     4 wushmachine  staff   128B Apr 24 10:08 .ipynb_checkpoints
-rw-r--r--@    1 wushmachine  staff   1.6M Apr 24 10:06 20021010_easy_ham.tar.bz2
-rw-r--r--@    1 wushmachine  staff   997K Apr 24 10:07 20021010_hard_ham.tar.bz2
-rw-r--r--@    1 wushmachine  staff   1.1M Apr 24 10:07 20021010_spam.tar.bz2
-rw-r--r--@    1 wushmachine  staff   589B Apr 24 10:06 Assignment4.ipynb
-rw-r--r--@    1 wushmachine  staff   7.9K Apr 24 10:08 IDSAI_Assignment4_SP4_22-23.ipynb
drwx--x--x@ 2553 wushmachine  staff    80K Oct 10  2002 easy_ham
drwx--x--x@  252 wushmachine  staff   7.9K Dec 16  2004 hard_ham
-rw-r--r--@    1 wushmachine  staff   7.5K Apr 24 10:04 na‹ve_bayes_intro.ipynb
-rw-r--r--@    1 wushmachine  staff    11K Apr 24 10:04 probability_intro.ipynb
drwxr-xr-x@  503 wushmachine  staff

### 1. Preprocessing: 

##### 1.1 Look at a few emails from easy_ham, hard_ham and spam. Do you think you would be able to classify the emails just by inspection? How do you think a succesful model can learn the difference between the different classes of emails?


In [ ]:
# Write your code for here for looking a few emails

It's easy to tell the difference between easy_ham and spam, it's sometimes hard to tell the difference between hard-ham and spam, hard-ham often take the shape of poorly written newsletters, which resemble some of the strange offers embedded in spam messages

Answer 1.1:

##### 1.2 Note that the email files contain a lot of extra information, besides the actual message. Ignore that for now and run on the entire text (in the optional part further down can experiment with filtering out the headers and footers). We don’t want to train and test on the same data (it might help to reflect on why if you don't recall). Split the spam and the ham datasets in a training set and a test set. (`hamtrain`, `spamtrain`, `hamtest`, and `spamtest`). Use only the easy_ham part as ham data for quesions 1 and 2.

In [163]:
# Write your code for here for splitting the data

ham = datasets.load_files('./easy')
#hard_ham = datasets.load_files('./hard')
spam = datasets.load_files('./spm')
data = datasets.load_files('./Data')
print(SpamTrain.keys())
print(SpamTrain.DESCR)

dict_keys(['data', 'filenames', 'target_names', 'target', 'DESCR'])
None


In [114]:
#ranomizing order, might not be necessary.


In [69]:
data['target']

array([0, 0, 0, ..., 0, 1, 1])

In [ ]:
divide the data into words

create a dictionary of words from the dataset, if the target is spam

In [168]:
coun_vect = CountVectorizer(decode_error = 'ignore')
count_matrix = coun_vect.fit_transform(data['data'])
count_array = count_matrix.toarray()
df = pd.DataFrame(data=count_array,columns = coun_vect.get_feature_names_out())
#print(df["money"])



3052

In [169]:

#split the data
data_train, data_test, target_train, target_test = train_test_split(df, data['target'], test_size=0.3) #,random_state=109)


In [172]:
gnb = GaussianNB()
#Train the model using the training sets
gnb.fit(data_train, target_train)

#Predict the response for test dataset
target_pred = gnb.predict(data_test)

In [173]:
print("Accuracy:",metrics.accuracy_score(target_test, target_pred))

Accuracy: 0.9705240174672489


In [146]:

ham_vectorizer = CountVectorizer(decode_error = 'ignore')
# tokenize and build vocab
ham_vectorizer.fit(hamlist)

print(ham_vectorizer.vocabulary_)

# encode document
ham_vector = ham_vectorizer.transform(hamlist)
# summarize encoded vector
#print(ham_vector.shape)
hamdict =(ham_vector.toarray())
hamdict

{'from': 19992, 'rssfeeds': 33821, 'jmason': 26013, 'org': 30139, 'wed': 39139, 'oct': 29846, '11': 2114, '43': 6279, '53': 6832, '2002': 3781, 'return': 33431, 'path': 30734, 'example': 18712, 'com': 14580, 'delivered': 16381, 'to': 37339, 'yyyy': 39940, 'localhost': 27228, 'received': 32797, 'jalapeno': 25876, '127': 2285, 'by': 13226, 'postfix': 31572, 'with': 39421, 'esmtp': 18543, 'id': 24758, '643dc16f03': 7419, 'for': 19740, 'jm': 26012, '0100': 797, 'ist': 25775, 'imap': 24979, 'fetchmail': 19353, 'single': 34998, 'drop': 17369, '02': 919, 'dogma': 17151, 'slashnull': 35104, 'g9282kk31781': 22248, '09': 1446, '46': 6393, 'message': 28215, '200210020802': 4376, 'guardian': 23732, 'subject': 36111, 'senators': 34502, 'threaten': 37147, 'rein': 33025, 'in': 25080, 'bush': 13181, 'date': 15974, '08': 1396, '0000': 2, 'content': 15035, 'type': 37851, 'text': 36983, 'plain': 31284, 'encoding': 18262, 'utf': 38403, 'spam': 35460, 'status': 35820, 'no': 29489, 'hits': 24322, '692': 759

array([[1, 0, 1, ..., 0, 0, 0],
       [0, 0, 5, ..., 5, 0, 7],
       [1, 0, 3, ..., 0, 0, 0],
       ...,
       [1, 1, 3, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 2, ..., 0, 0, 0]])

In [143]:
spam_vectorizer = CountVectorizer(decode_error = 'ignore')
# tokenize and build vocab
spam_vectorizer.fit(spamlist)

print(spam_vectorizer.vocabulary_)

# encode document
spam_vector = spam_vectorizer.transform(spamlist)
# summarize encoded vector
print(spam_vector.shape)
spamdict =(spam_vector.toarray())

{'from': 13022, 'allcontacts': 6526, 'bounce': 8096, 'briefs': 8186, 'ein': 11625, 'cz': 9996, 'thu': 24787, 'sep': 23159, '11': 1333, '46': 3476, '23': 2231, '2002': 1908, 'return': 22291, 'path': 20043, 'delivered': 10349, 'to': 24915, 'zzzz': 28481, 'localhost': 17325, 'jmason': 16063, 'org': 19646, 'received': 21936, 'jalapeno': 15868, '127': 1417, 'by': 8371, 'zzzzason': 28486, 'postfix': 20757, 'with': 27124, 'esmtp': 11998, 'id': 14919, '120d616fd6': 1394, 'for': 12854, '35': 2829, '0100': 805, 'ist': 15669, 'imap': 15184, 'fetchmail': 12566, 'single': 23425, 'drop': 11109, '05': 958, 'bay': 7562, '62': 4178, '24': 2275, '69': 4285, '193': 1693, 'dogma': 10803, 'slashnull': 23494, 'g84lt3z19745': 13294, 'webmaster': 26951, 'efi': 11577, 'ie': 14967, 'wed': 26964, '22': 2161, '29': 2445, '03': 917, 'oak': 19323, '194': 1696, '60ec413923': 4163, '16': 1543, '51': 3774, '49': 3550, '0200': 887, 'cest': 8861, 'userid': 25828, '1002': 1264, 'ceeb11bed5': 8813, '28': 2410, 'media': 17

In [144]:
hamspam_vectorizer = CountVectorizer(decode_error = 'ignore')
# tokenize and build vocab
hamspam_vectorizer.fit_transform(hamspamlist)

print(hamspam_vectorizer.vocabulary_)

# encode document
hamspam_vector = hamspam_vectorizer.transform(hamspamlist)
# summarize encoded vector
print(hamspam_vector.shape)
hamspamdict = (hamspam_vector.toarray())

{'from': 29912, 'rssfeeds': 50687, 'jmason': 38230, 'org': 44994, 'wed': 59015, 'oct': 44523, '11': 3376, '43': 9194, '53': 10098, '2002': 5461, 'return': 50051, 'path': 45869, 'example': 28108, 'com': 22118, 'delivered': 24496, 'to': 55624, 'yyyy': 60727, 'localhost': 40450, 'received': 49281, 'jalapeno': 37929, '127': 3605, 'by': 20080, 'postfix': 47220, 'with': 59376, 'esmtp': 27858, 'id': 36164, '643dc16f03': 11019, 'for': 29571, 'jm': 38224, '0100': 1590, 'ist': 37641, 'imap': 36605, 'fetchmail': 29000, 'single': 52359, 'drop': 26013, '02': 1791, 'dogma': 25540, 'slashnull': 52509, 'g9282kk31781': 32608, '09': 2441, '46': 9370, 'message': 41914, '200210020802': 6224, 'guardian': 34401, 'subject': 53823, 'senators': 51718, 'threaten': 55329, 'rein': 49562, 'in': 36723, 'bush': 19992, 'date': 23980, '08': 2376, '0000': 2, 'content': 22656, 'type': 56365, 'text': 55110, 'plain': 46807, 'encoding': 27453, 'utf': 57369, 'spam': 52999, 'status': 53458, 'no': 43918, 'hits': 35294, '692':

(2136, 61380)


In [ ]:
for k in ham_vector.keys():
    denom = hamspam

In [94]:
#Create a Naïve Bayes Classifier
gnb = GaussianNB()

#Train the model using the training sets
gnb.fit(data_train, target_train)

#Predict the response for test dataset
y_pred = gnb.predict(data_test)

ValueError: Expected 2D array, got 1D array instead:
array=[b"From rssfeeds@jmason.org  Wed Oct  2 11:43:53 2002\nReturn-Path: <rssfeeds@example.com>\nDelivered-To: yyyy@localhost.example.com\nReceived: from localhost (jalapeno [127.0.0.1])\n\tby jmason.org (Postfix) with ESMTP id 643DC16F03\n\tfor <jm@localhost>; Wed,  2 Oct 2002 11:43:53 +0100 (IST)\nReceived: from jalapeno [127.0.0.1]\n\tby localhost with IMAP (fetchmail-5.9.0)\n\tfor jm@localhost (single-drop); Wed, 02 Oct 2002 11:43:53 +0100 (IST)\nReceived: from dogma.slashnull.org (localhost [127.0.0.1]) by\n    dogma.slashnull.org (8.11.6/8.11.6) with ESMTP id g9282kK31781 for\n    <jm@jmason.org>; Wed, 2 Oct 2002 09:02:46 +0100\nMessage-Id: <200210020802.g9282kK31781@dogma.slashnull.org>\nTo: yyyy@example.com\nFrom: guardian <rssfeeds@example.com>\nSubject: Senators threaten to rein in Bush\nDate: Wed, 02 Oct 2002 08:02:46 -0000\nContent-Type: text/plain; encoding=utf-8\nX-Spam-Status: No, hits=-692.4 required=5.0\n\ttests=AWL\n\tversion=2.50-cvs\nX-Spam-Level: \n\nURL: http://www.newsisfree.com/click/-3,8456087,215/\nDate: 2002-10-02T03:56:58+01:00\n\n*World latest:* Leading Republicans side with the Democrats over the \npresident's war powers.\n\n\n"
 b"From Stewart.Smith@ee.ed.ac.uk  Thu Aug 22 14:44:26 2002\nReturn-Path: <Stewart.Smith@ee.ed.ac.uk>\nDelivered-To: zzzz@localhost.netnoteinc.com\nReceived: from localhost (localhost [127.0.0.1])\n\tby phobos.labs.netnoteinc.com (Postfix) with ESMTP id EC69D47C66\n\tfor <zzzz@localhost>; Thu, 22 Aug 2002 09:44:25 -0400 (EDT)\nReceived: from phobos [127.0.0.1]\n\tby localhost with IMAP (fetchmail-5.9.0)\n\tfor zzzz@localhost (single-drop); Thu, 22 Aug 2002 14:44:25 +0100 (IST)\nReceived: from n6.grp.scd.yahoo.com (n6.grp.scd.yahoo.com [66.218.66.90])\n    by dogma.slashnull.org (8.11.6/8.11.6) with SMTP id g7MDcOZ08504 for\n    <zzzz@example.com>; Thu, 22 Aug 2002 14:38:25 +0100\nX-Egroups-Return: sentto-2242572-52736-1030023506-zzzz=example.com@returns.groups.yahoo.com\nReceived: from [66.218.67.192] by n6.grp.scd.yahoo.com with NNFMP;\n    22 Aug 2002 13:38:26 -0000\nX-Sender: Stewart.Smith@ee.ed.ac.uk\nX-Apparently-To: zzzzteana@yahoogroups.com\nReceived: (EGP: mail-8_1_0_1); 22 Aug 2002 13:38:25 -0000\nReceived: (qmail 48882 invoked from network); 22 Aug 2002 13:38:25 -0000\nReceived: from unknown (66.218.66.218) by m10.grp.scd.yahoo.com with QMQP;\n    22 Aug 2002 13:38:25 -0000\nReceived: from unknown (HELO postbox.ee.ed.ac.uk) (129.215.80.253) by\n    mta3.grp.scd.yahoo.com with SMTP; 22 Aug 2002 13:38:24 -0000\nReceived: from ee.ed.ac.uk (sxs@dunblane [129.215.34.86]) by\n    postbox.ee.ed.ac.uk (8.11.0/8.11.0) with ESMTP id g7MDcNi28645 for\n    <forteana@yahoogroups.com>; Thu, 22 Aug 2002 14:38:23 +0100 (BST)\nMessage-Id: <3D64E94E.8060301@ee.ed.ac.uk>\nOrganization: Scottish Microelectronics Centre\nUser-Agent: Mozilla/5.0 (X11; U; SunOS sun4u; en-US; rv:1.1b) Gecko/20020628\nX-Accept-Language: en, en-us\nTo: zzzzteana@yahoogroups.com\nReferences: <3D64F325.11319.61EA648@localhost>\nFrom: Stewart Smith <Stewart.Smith@ee.ed.ac.uk>\nX-Yahoo-Profile: stochasticus\nMIME-Version: 1.0\nMailing-List: list zzzzteana@yahoogroups.com; contact\n    forteana-owner@yahoogroups.com\nDelivered-To: mailing list zzzzteana@yahoogroups.com\nPrecedence: bulk\nList-Unsubscribe: <mailto:zzzzteana-unsubscribe@yahoogroups.com>\nDate: Thu, 22 Aug 2002 14:38:22 +0100\nSubject: Re: [zzzzteana] Nothing like mama used to make\nReply-To: zzzzteana@yahoogroups.com\nContent-Type: text/plain; charset=US-ASCII\nContent-Transfer-Encoding: 7bit\n\n>  in adding cream to spaghetti carbonara, which has the same effect on pasta as\n>  making a pizza a deep-pie; \n\nI just had to jump in here as Carbonara is one of my favourites to make and ask \nwhat the hell are you supposed to use instead of cream?  I've never seen a \nrecipe that hasn't used this.  Personally I use low fat creme fraiche because it \nworks quite nicely but the only time I've seen an supposedly authentic recipe \nfor carbonara  it was identical to mine (cream, eggs and lots of fresh parmesan) \nexcept for the creme fraiche.\n\nStew\n-- \nStewart Smith\nScottish Microelectronics Centre, University of Edinburgh.\nhttp://www.ee.ed.ac.uk/~sxs/\n\n\n------------------------ Yahoo! Groups Sponsor ---------------------~-->\n4 DVDs Free +s&p Join Now\nhttp://us.click.yahoo.com/pt6YBB/NXiEAA/mG3HAA/7gSolB/TM\n---------------------------------------------------------------------~->\n\nTo unsubscribe from this group, send an email to:\nforteana-unsubscribe@egroups.com\n\n \n\nYour use of Yahoo! Groups is subject to http://docs.yahoo.com/info/terms/ \n\n\n\n"
 b'From johnhall@evergo.net  Tue Sep 17 23:29:48 2002\nReturn-Path: <johnhall@evergo.net>\nDelivered-To: yyyy@localhost.example.com\nReceived: from localhost (jalapeno [127.0.0.1])\n\tby jmason.org (Postfix) with ESMTP id CF5C316F03\n\tfor <jm@localhost>; Tue, 17 Sep 2002 23:29:47 +0100 (IST)\nReceived: from jalapeno [127.0.0.1]\n\tby localhost with IMAP (fetchmail-5.9.0)\n\tfor jm@localhost (single-drop); Tue, 17 Sep 2002 23:29:47 +0100 (IST)\nReceived: from mail.evergo.net ([206.191.151.2]) by dogma.slashnull.org\n    (8.11.6/8.11.6) with SMTP id g8HKUvC25665 for <jm@jmason.org>;\n    Tue, 17 Sep 2002 21:30:58 +0100\nReceived: (qmail 31515 invoked from network); 17 Sep 2002 20:31:20 -0000\nReceived: from dsl.206.191.151.102.evergo.net (HELO JMHALL)\n    (206.191.151.102) by mail.evergo.net with SMTP; 17 Sep 2002 20:31:20 -0000\nReply-To: <johnhall@evergo.net>\nFrom: "John Hall" <johnhall@evergo.net>\nTo: <yyyy@example.com>, "\'Gary Lawrence Murphy\'" <garym@canada.com>\nCc: "\'Stephen D. Williams\'" <sdw@lig.net>, <fork@example.com>,\n\t<lea@lig.net>\nSubject: RE: Slaughter in the Name of God\nDate: Tue, 17 Sep 2002 13:31:20 -0700\nMessage-Id: <001601c25e89$2f06a3d0$0200a8c0@JMHALL>\nMIME-Version: 1.0\nContent-Type: text/plain; charset="us-ascii"\nContent-Transfer-Encoding: 7bit\nX-Priority: 3 (Normal)\nX-Msmail-Priority: Normal\nX-Mailer: Microsoft Outlook, Build 10.0.2627\nIn-Reply-To: <20020917165028.4F4EA16F03@example.com>\nImportance: Normal\nX-Mimeole: Produced By Microsoft MimeOLE V6.00.2600.0000\nX-Spam-Status: No, hits=-6.0 required=7.0\n\ttests=AWL,IN_REP_TO,QUOTED_EMAIL_TEXT\n\tversion=2.50-cvs\nX-Spam-Level: \n\n\n\n> From: yyyy@example.com [mailto:yyyy@example.com]\n> Sent: Tuesday, September 17, 2002 9:50 AM\n> > ... we\'ve been fighting the War on Terrorism for as long\n> > as there\'s been commerce, so you\'d think we\'d /realize/ that\n> > escalation of violence is not a solution.\n> \n> Well said!\n> \n> --j.\n\nYeah.  It certainly wasn\'t a solution to the Carthaginian problem or the\nBarbary Pirates.  Wait ... no ... actually ... it was a rather permanent\nsolution.\n\n\n'
 ...
 b'From fork-admin@xent.com  Fri Aug 23 11:08:31 2002\nReturn-Path: <fork-admin@xent.com>\nDelivered-To: yyyy@localhost.netnoteinc.com\nReceived: from localhost (localhost [127.0.0.1])\n\tby phobos.labs.netnoteinc.com (Postfix) with ESMTP id BE9C147C69\n\tfor <jm@localhost>; Fri, 23 Aug 2002 06:06:48 -0400 (EDT)\nReceived: from phobos [127.0.0.1]\n\tby localhost with IMAP (fetchmail-5.9.0)\n\tfor jm@localhost (single-drop); Fri, 23 Aug 2002 11:06:48 +0100 (IST)\nReceived: from xent.com ([64.161.22.236]) by dogma.slashnull.org\n    (8.11.6/8.11.6) with ESMTP id g7MK4SZ22618 for <jm@jmason.org>;\n    Thu, 22 Aug 2002 21:04:29 +0100\nReceived: from lair.xent.com (localhost [127.0.0.1]) by xent.com (Postfix)\n    with ESMTP id E6AD62940D8; Thu, 22 Aug 2002 13:02:08 -0700 (PDT)\nDelivered-To: fork@example.com\nReceived: from mail.evergo.net (unknown [206.191.151.2]) by xent.com\n    (Postfix) with SMTP id 790A7294099 for <fork@xent.com>; Thu,\n    22 Aug 2002 13:01:08 -0700 (PDT)\nReceived: (qmail 21546 invoked from network); 22 Aug 2002 20:02:51 -0000\nReceived: from dsl.206.191.151.102.evergo.net (HELO JMHALL)\n    (206.191.151.102) by mail.evergo.net with SMTP; 22 Aug 2002 20:02:51 -0000\nReply-To: <johnhall@evergo.net>\nFrom: "John Hall" <johnhall@evergo.net>\nTo: <fork@example.com>\nSubject: Property rights in the 3rd World (De Soto\'s Mystery of Capital)\nMessage-Id: <008701c24a16$e3443830$0200a8c0@JMHALL>\nMIME-Version: 1.0\nContent-Type: text/plain; charset="US-ASCII"\nContent-Transfer-Encoding: 7bit\nX-Priority: 3 (Normal)\nX-Msmail-Priority: Normal\nX-Mailer: Microsoft Outlook, Build 10.0.2627\nIn-Reply-To: <200208221811.LAA21283@maltesecat>\nX-Mimeole: Produced By Microsoft MimeOLE V6.00.2600.0000\nImportance: Normal\nSender: fork-admin@xent.com\nErrors-To: fork-admin@xent.com\nX-Beenthere: fork@example.com\nX-Mailman-Version: 2.0.11\nPrecedence: bulk\nList-Help: <mailto:fork-request@xent.com?subject=help>\nList-Post: <mailto:fork@example.com>\nList-Subscribe: <http://xent.com/mailman/listinfo/fork>, <mailto:fork-request@xent.com?subject=subscribe>\nList-Id: Friends of Rohit Khare <fork.xent.com>\nList-Unsubscribe: <http://xent.com/mailman/listinfo/fork>,\n    <mailto:fork-request@xent.com?subject=unsubscribe>\nList-Archive: <http://xent.com/pipermail/fork/>\nDate: Thu, 22 Aug 2002 13:02:45 -0700\nX-Pyzor: Reported 0 times.\nX-Spam-Status: No, hits=-7.0 required=7.0\n\ttests=IN_REP_TO,KNOWN_MAILING_LIST,QUOTED_EMAIL_TEXT,\n\t      SPAM_PHRASE_01_02\n\tversion=2.40-cvs\nX-Spam-Level: \n\n\n\n> From: fork-admin@xent.com [mailto:fork-admin@xent.com] On Behalf Of\nDave\n> Long\n> Sent: Thursday, August 22, 2002 11:12 AM\n> To: fork@example.com\n> Subject: RE: The Curse of India\'s Socialism\n> \n> \n> \n> When I\'d read that "getting legal title\n> can take 20 years", when I believe that\n> 1 year ought to be more than sufficient,\n> (and helped by the Cairo reference) I\'d\n> assumed that we were talking about the\n> urban poor.\n> \n> If I see people living in mansions, or\n> even in suburban subdivisions, I assume\n> they didn\'t have too much trouble with\n> their titles.\n\nPg 177:\nIn another country, a local newspaper, intrigued by our evidence of\nextralegal real estate holdings, checked to see if the head of state\'s\nofficial residence had a recorded title.  It did not.\n\nPg 92:\nThe value of land in the formal sector of Lima averages US$50 per square\nmeter, whereas in the area of Gamarra, where a great deal of Peru\'s\ninformal manufacturing sector resides, the value per square meter can go\nas high as US$3,000.\n\n==========\n\nI\'d have made the same assumption you did.  De Soto says that isn\'t\ncorrect.  You can find mansions that don\'t have title.  A lot of them,\nin fact.  But they can\'t be used for collateral for a loan, or otherwise\nparticipate as \'capital\' because of their extra-legal status.\n\n \n> > Mr. Long, I think you\'d particularly enjoy the De Soto work.\n> \n> On the "to find" list.  Any chance of\n> an explanation of that "Bell Jar" in\n> the meantime?\n\nFrench historian Fernand Braudel (so Braudel\'s Bell Jar, not De Soto\'s)\n\n==>\n\nThe key problem is to find out why that sector of society of the past,\nwhich I would not hesitate to call capitalist, should have lived as if\nin a bell jar, cut off from the rest; why was it not able to expand and\nconquer the whole of society? ... [Why was it that] a significant rate\nof capital formation was possible only in certain sectors and not in the\nwhole market economy of the time? ... It would perhaps be teasingly\nparadoxial to say that whatever was in short supply, money certainly was\nnot ... so this was an age where poor land was bought up and magnificent\ncountry residents built ... [How do we] resolve the contradiction ...\nbetween the depressed economic climate and the splendors of Florence\nunder Lorenzo the Magnificent?\n \n\n--------------\n\nDe Soto\'s theory is that the Bell Jar is formed when you segregate those\nwho have *practical* access to legal property rights and those who do\nnot.  The poor[1] have property -- lots and lots of property.  What they\ndon\'t have is access to the systems where we turn property into capital\nand allow it to start growing.  Their property can only be exchanged\nwith a small section of people who know them personally.\n\n[1] Actual poor people, not \'poor\' Americans with a living standard that\nis the envy of most of the world.\n\n\nhttp://xent.com/mailman/listinfo/fork\n\n'
 b'From fork-admin@xent.com  Fri Sep 20 11:32:42 2002\nReturn-Path: <fork-admin@xent.com>\nDelivered-To: yyyy@localhost.example.com\nReceived: from localhost (jalapeno [127.0.0.1])\n\tby jmason.org (Postfix) with ESMTP id CA74016F03\n\tfor <jm@localhost>; Fri, 20 Sep 2002 11:32:41 +0100 (IST)\nReceived: from jalapeno [127.0.0.1]\n\tby localhost with IMAP (fetchmail-5.9.0)\n\tfor jm@localhost (single-drop); Fri, 20 Sep 2002 11:32:41 +0100 (IST)\nReceived: from xent.com ([64.161.22.236]) by dogma.slashnull.org\n    (8.11.6/8.11.6) with ESMTP id g8JLlhC05097 for <jm@jmason.org>;\n    Thu, 19 Sep 2002 22:47:44 +0100\nReceived: from lair.xent.com (localhost [127.0.0.1]) by xent.com (Postfix)\n    with ESMTP id F0A9029416D; Thu, 19 Sep 2002 14:44:05 -0700 (PDT)\nDelivered-To: fork@example.com\nReceived: from hotmail.com (dav56.law15.hotmail.com [64.4.22.191]) by\n    xent.com (Postfix) with ESMTP id 650A729409C for <fork@xent.com>;\n    Thu, 19 Sep 2002 14:43:56 -0700 (PDT)\nReceived: from mail pickup service by hotmail.com with Microsoft SMTPSVC;\n    Thu, 19 Sep 2002 14:47:21 -0700\nX-Originating-Ip: [207.202.171.254]\nFrom: "Mr. FoRK" <fork_list@hotmail.com>\nTo: <fork@example.com>\nReferences: <20020919171804.GA18065@canonical.org>\nSubject: Re: Avast there matey\nMIME-Version: 1.0\nContent-Type: text/plain; charset="iso-8859-1"\nContent-Transfer-Encoding: 7bit\nX-Priority: 3\nX-Msmail-Priority: Normal\nX-Mailer: Microsoft Outlook Express 5.50.4133.2400\nX-Mimeole: Produced By Microsoft MimeOLE V5.50.4133.2400\nMessage-Id: <DAV56idcDjoYAPNXRPm00000075@hotmail.com>\nX-Originalarrivaltime: 19 Sep 2002 21:47:21.0730 (UTC) FILETIME=[221F5E20:01C26026]\nSender: fork-admin@xent.com\nErrors-To: fork-admin@xent.com\nX-Beenthere: fork@example.com\nX-Mailman-Version: 2.0.11\nPrecedence: bulk\nList-Help: <mailto:fork-request@xent.com?subject=help>\nList-Post: <mailto:fork@example.com>\nList-Subscribe: <http://xent.com/mailman/listinfo/fork>, <mailto:fork-request@xent.com?subject=subscribe>\nList-Id: Friends of Rohit Khare <fork.xent.com>\nList-Unsubscribe: <http://xent.com/mailman/listinfo/fork>,\n    <mailto:fork-request@xent.com?subject=unsubscribe>\nList-Archive: <http://xent.com/pipermail/fork/>\nDate: Thu, 19 Sep 2002 14:51:59 -0700\nX-Spam-Status: No, hits=-3.3 required=5.0\n\ttests=AWL,KNOWN_MAILING_LIST,QUOTED_EMAIL_TEXT,REFERENCES\n\tversion=2.50-cvs\nX-Spam-Level: \n\n\n----- Original Message -----\nFrom: "Kragen Sitaker" <kragen@pobox.com>\n\n> > It\'s an exclamation that may allude to a ship striking some rock or\n> > other obstacle so hard that her timbers shiver, or shake, so implying a\n> > calamity has occurred. It is first recorded as being used by Captain\n> > Frederick Marryat in /Jacob Faithful/ in 1835: "I won\'t thrash you Tom.\n> > Shiver my timbers if I do".\nI think it went like this: "I won\'t thrash you Tom, if you shiver my\ntimber..."\nOr maybe that was just the butt pirates...\n\n\n'
 b'From fork-admin@xent.com  Thu Sep 26 11:04:50 2002\nReturn-Path: <fork-admin@xent.com>\nDelivered-To: yyyy@localhost.example.com\nReceived: from localhost (jalapeno [127.0.0.1])\n\tby jmason.org (Postfix) with ESMTP id 9A72816F03\n\tfor <jm@localhost>; Thu, 26 Sep 2002 11:04:49 +0100 (IST)\nReceived: from jalapeno [127.0.0.1]\n\tby localhost with IMAP (fetchmail-5.9.0)\n\tfor jm@localhost (single-drop); Thu, 26 Sep 2002 11:04:49 +0100 (IST)\nReceived: from xent.com ([64.161.22.236]) by dogma.slashnull.org\n    (8.11.6/8.11.6) with ESMTP id g8PMxPC17679 for <jm@jmason.org>;\n    Wed, 25 Sep 2002 23:59:26 +0100\nReceived: from lair.xent.com (localhost [127.0.0.1]) by xent.com (Postfix)\n    with ESMTP id 7AB48294164; Wed, 25 Sep 2002 15:55:09 -0700 (PDT)\nDelivered-To: fork@example.com\nReceived: from jamesr.best.vwh.net (jamesr.best.vwh.net [192.220.76.165])\n    by xent.com (Postfix) with SMTP id AB87D29409A for <fork@xent.com>;\n    Wed, 25 Sep 2002 15:54:07 -0700 (PDT)\nReceived: (qmail 7297 invoked by uid 19621); 25 Sep 2002 22:55:59 -0000\nReceived: from unknown (HELO avalon) ([64.125.200.18]) (envelope-sender\n    <jamesr@best.com>) by 192.220.76.165 (qmail-ldap-1.03) with SMTP for\n    <fork@xent.com>; 25 Sep 2002 22:55:59 -0000\nSubject: Re: Digital radio playlists are prohibited?!\nFrom: James Rogers <jamesr@best.com>\nTo: fork@example.com\nIn-Reply-To: <132203498905.20020925163453@magnesium.net>\nReferences: <95745878-D05F-11D6-8F1E-000393A46DEA@alumni.caltech.edu>\n    <132203498905.20020925163453@magnesium.net>\nContent-Type: text/plain\nContent-Transfer-Encoding: 7bit\nX-Mailer: Evolution/1.0.2-5mdk\nMessage-Id: <1032995753.27386.88.camel@avalon>\nMIME-Version: 1.0\nSender: fork-admin@xent.com\nErrors-To: fork-admin@xent.com\nX-Beenthere: fork@example.com\nX-Mailman-Version: 2.0.11\nPrecedence: bulk\nList-Help: <mailto:fork-request@xent.com?subject=help>\nList-Post: <mailto:fork@example.com>\nList-Subscribe: <http://xent.com/mailman/listinfo/fork>, <mailto:fork-request@xent.com?subject=subscribe>\nList-Id: Friends of Rohit Khare <fork.xent.com>\nList-Unsubscribe: <http://xent.com/mailman/listinfo/fork>,\n    <mailto:fork-request@xent.com?subject=unsubscribe>\nList-Archive: <http://xent.com/pipermail/fork/>\nDate: 25 Sep 2002 16:15:52 -0700\nX-Spam-Status: No, hits=-4.6 required=5.0\n\ttests=AWL,EMAIL_ATTRIBUTION,IN_REP_TO,KNOWN_MAILING_LIST,\n\t      PLING_QUERY,QUOTED_EMAIL_TEXT,REFERENCES,REPLY_WITH_QUOTES\n\tversion=2.50-cvs\nX-Spam-Level: \n\nOn Wed, 2002-09-25 at 13:34, bitbitch@magnesium.net wrote:\n> \n> This, kiddies was apparently the legislative beginnings of the whole\n> streaming audio-gets-spanked-by-fees ruling that came down in the\n> earlier parts of this year.  This first act applied to non-exempt,\n> non-subscription transmission services.  When Congress got around in\n> 1998 and realized that webcasting services -might- be different\n> (though I honestly can\'t see how) they wrote in the provision through\n> the DMCA to include such transmissions.\n\n\nThe restrictive law regarding audio is actually the accumulated cruft of\n30 years of various legislative acts. The totality of what we have now\ncome from various parts of all the following re: sound recordings:\n\n1998 - DMCA\n1995 - Digital Performance Right in Sound Recordings Act\n1992 - Audio Home Recording Act\n1976 - Copyright Act amendment\n1972 - Copyright Act amendment\n\nIt is worth noting that many people have forgotten about the 1976\nCopyright Act Amendment which created the foundational law stating that\nthe copyright owners have the right to limit personal use of audio\nrecordings after First Sale even if you are not "making copies" in any\ncommercial sense.  Sound recordings, for many intents and purposes, are\nexplicitly excluded from Fair Use by the 1976 amendment.\n\n-James Rogers\n jamesr@best.com\n\n\n\n'].
Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.

### 2.1 Write a Python program that: 
1.	Uses the four datasets from Question 1 (`hamtrain`, `spamtrain`, `hamtest`, and `spamtest`) 
2.	Trains a Naïve Bayes classifier (use the [scikit-learn library](https://scikit-learn.org/stable/)) on `hamtrain` and `spamtrain`, that classifies the test sets and reports True Positive and False Negative rates on the `hamtest` and `spamtest` datasets. Use `CountVectorizer` ([Documentation here](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html#sklearn.feature_extraction.text.CountVectorizer)) to transform the email texts into vectors. Please note that there are different types of Naïve Bayes Classifier in scikit-learn ([Documentation here](https://scikit-learn.org/stable/modules/naive_bayes.html)). Test two of these classifiers that are well suited for this problem:
- Multinomial Naive Bayes  
- Bernoulli Naive Bayes. 

Please inspect the documentation to ensure input to the classifiers is appropriate before you start coding. 



In [ ]:
# Write your code here

### 2.2 Answer the following questions:
##### a) What does the CountVectorizer do?
Answer 2.2.a
##### b) What is the difference between Multinomial Naive Bayes and Bernoulli Naive Bayes
Answer 2.2.b


### 3.1 Run the two models:
Run (don't retrain) the two models from Question 2 on spam versus hard-ham. Does the performance differ compared to question 2 when the model was run on spam versus easy-ham? If so, why?

In [ ]:
# Write your code here

Answer 3.1:

### 3.2 Retrain
Retrain new Multinomial and Bernolli Naive Bayes classifers on the combined (easy+hard) ham and spam. Now evaluate on spam versus hard-ham as in 3.1. Also evaluate on spam versus easy-ham. Compare the performance with question 2 and 3.1. What do you observe?

In [ ]:
# Write your code here

Answer 3.2:

### 3.3 Further improvements
Do you have any suggestions for how performance could be further improved? You don't have to implement them, just present your ideas.

Answer 3.3: